### DISCUSSION TITLE
Semester 4 Week 3

### DISCUSSION THREAD

**Discussion by LPkkjHD** [SE-TINF22B6/Underwatch#160] [Announcements]
Hello and welcome to this weeks update on what has happened in Project Underwatch.

### Software Architecture Documentation (SAD)

We spend this week on expanding the SAD template Mrs. Li provided to us with the first diagrams showcasing how we are going to tacke our project architecturally. Of course this is more of a documentation of the status quo than a design blueprint as we have already implemented quite a lot of the system already - as you've probably seen in our past week's reports.

If you want to check out the document in detail, we welcome you to check it out [here](https://github.com/SE-TINF22B6/Underwatch/blob/main/documentation/sad.md).
First off we want to showcase a high level overview of what systems we are building as well as how they communicate with each other - assuming normal operation:

![High Level](https://raw.githubusercontent.com/SE-TINF22B6/Underwatch/main/documentation/assets/ComponentDiagram.drawio.svg)

As you can see we do have three subsystems which interact with each other via a `RESTful` API.
This also brings us to one of the employed architecture patterns we have decided on - with a data storage being available via a `RESTful` application we decouple the game (which is replicated for every player playing concurrently) from the Scoreboard which acts on the web page.

The sequence diagram below depics how we are going to structure the game loop. The game loop is a loop which is steadily executed about 60 times per second (or however often you set your FPS to). 

![Sequence Diagram Gameloop](https://raw.githubusercontent.com/SE-TINF22B6/Underwatch/main/documentation/assets/sadSequenceDiagram.svg)

### 🗺️ Web Page progress

On the web page we spend a lot of effort into improving the general user experience when interacting with the page (UX), resulting in a full blown refactoring of the Scoreboard page, making it more performant due to less code being delivered to the user.

### 🎮 Game progress

In the game there has been a lot of progress on multiple fronts, which is more than we can cover here - you're more than welcome to test out this release and see the changes for yourself.

- **Better animations**: We have replaced the filler animations and assets for the player with the ones we are planning to use in the final release of the game, also adding in animations for walking and idling into any of the four directions `LEFT, RIGHT, UP, DOWN`.

https://github.com/SE-TINF22B6/Underwatch/assets/11832982/87277873-c2c6-4c4d-b3e5-35a701edf330

- **Weapons**: We worked on the weapon system and how we want the current weapon to be displayed in the game and settled on a weapon, which rotates around the player into the direction of the crosshair. This is still a WIP feature, but we are more than happy to share this progress with you.

https://github.com/SE-TINF22B6/Underwatch/assets/11832982/150ad098-7783-4d60-b218-9e8610c770f2

As per usual, it has been a pleasure to update you to the most recent state of this project. We hoped you enjoyed reading this article - if you have any questions or comments let us know down below.

~ Team Underwatch

**Comment by sfrancisw** [comment]
Dear underwatch team.
Great progress. It is good that you work on your game and your documentation side by side. 
The new animations look really good. We're excited to see what you will work on next week.

Your team t2s

**Comment by HowDoIUseGit** [comment]
Moin Underwatch,
The expansion of the SAD template, along with clear diagrams showcasing architectural plans, demonstrates thoughtful documentation and planning. Decoupling the game from the Scoreboard through a RESTful API shows smart architectural decisions. The improvements to the web page's user experience and the advancements in the game, such as better animations and a dynamic weapon display, highlight your team's dedication to quality.
Keep up the good work!
Your CookPit

**Comment by fehlend0** [comment]
Dear Underwatch-Team, 

it's great to see the progress you made in your project! The expansion of the documentation and the focus on improving both the web page and game experience show a commitment to quality and user satisfaction. 

The detailed updates on system architecture and game development provide a very good insight into the project's direction and accomplishments. 

Keep up the excellent work!

Your Statify-Team

**Comment by Shiny1708** [comment]
The documentation you provided is very thorough and helps to easily understand the underlying architecture, great work there! Choosing a restful API to exchange information makes a lot of sense and is the right decision in my opinion. Having seen the progress on the website in person I can also just give a big thumbs up. Thing are looking more and more polished and look almost ready for release
